internal/controller/grafanaorganization_controller.go

```diff
@@ -99,17 +101,79 @@ func (r GrafanaOrganizationReconciler) reconcileCreate(ctx context.Context, graf
+	if grafanaOrganization.Status.OrgID == "" {
+		return ctrl.Result{}, r.createOrganizationInGrafana(ctx, grafanaAPI, grafanaOrganization)
+	} else {
+		// Cast orgID to int64
+		orgID, err := strconv.ParseInt(grafanaOrganization.Status.OrgID, 10, 64)
+		if err != nil {
+			logger.Error(err, "Failed to parse OrgID")
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+
+		grievious, err := grafanaAPI.Orgs.GetOrgByID(orgID)
+		if err != nil {
+			logger.Error(err, "Failed to get organization by ID")
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+
+		if grievious != nil {
+			// If the CR orgID matches an existing org in grafana, check if the name is the same as the CR
+			if grievious.Payload.Name != grafanaOrganization.Name {
+				// if the name of the CR is different from the name of the org in Grafana, update the name of the org in Grafana using the CR's display name.
+				_, err := grafanaAPI.Orgs.UpdateOrg(orgID, &grafanaAPIModels.UpdateOrgForm{
+					Name: grafanaOrganization.Spec.DisplayName,
+				})
+				if err != nil {
+					logger.Error(err, "Failed to update organization name")
+					return ctrl.Result{}, errors.WithStack(err)
+				}
+			} else {
+				return ctrl.Result{}, errors.Errorf("A grafana organization with the same name and ID already exists")
+			}
+		} else { // If the granfana organization CR has an orgID  but does not exist in Grafana, create the organization
+			return ctrl.Result{}, r.createOrganizationInGrafana(ctx, grafanaAPI, grafanaOrganization)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r GrafanaOrganizationReconciler) createOrganizationInGrafana(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, grafanaOrganization *v1alpha1.GrafanaOrganization) error {
+	logger := log.FromContext(ctx)
+
+	// Check if the organization name is available
+	obiwan, err := grafanaAPI.Orgs.GetOrgByName(grafanaOrganization.Name)
 	if err != nil {
-		logger.Error(err, "Could not rename Main Org. to Shared Org.")
-		return ctrl.Result{}, errors.WithStack(err)
+		logger.Error(err, "Failed to check if organization name is available")
+		return errors.WithStack(err)
 	}
+	// If the organization name is already taken, return an error
+	if obiwan != nil {
+		return errors.Errorf("Organization name is already taken")
+	} else {
+		logger.Info("Create organization in Grafana")
+
+		// If the name is available, create the organization in Grafana
+		createdOrg, err := grafanaAPI.Orgs.CreateOrg(&grafanaAPIModels.CreateOrgCommand{
+			Name: grafanaOrganization.Name,
+		})
+		if err != nil {
+			logger.Error(err, "Organization failed")
+			return errors.WithStack(err)
+		}
 
-	//TODO Implement the logic to create the Grafana organization
+		// Update the grafanaOrganization status with the orgID
+		grafanaOrganization.Status.OrgID = fmt.Sprintf("%d", createdOrg.Payload.OrgID)
+		if err = r.Status().Update(ctx, grafanaOrganization); err != nil {
```

Comment:
Maybe we can patch?

Reply:
That's what I did at first but then I saw that the best practices for updating a CR status from an operator was this one.

Reply:
Ack